### PR TITLE
Add detached spans API

### LINKED ISSUE #56: Split the logic of "current" span/context into a separate class
I would suggest to decouple the logic of "current" Context/Span from the Tracer class, by having a separate Context class similar with [io.grpc.Context](https://grpc.io/grpc-java/javadoc/io/grpc/Context.html)

Here is a possible interface, stolen from Java, that will also work for async (it is essentially a map key->value):

```
class Context {

class Key<V> {
// Gets the value from the Context.current() context for this key.
V get();

// Get the value from the specified context for this key.
V get(Context c);
}

// Creates a new context with the given key value set.
Context withValue(Context.Key<V> k1, V v1);

// Attaches this context, thus enter a new scope within which this context is current().
// The previously current context is returned.
Context attach();

// Reverses an attach(), restoring the previous context and exiting the current scope.
void detach(Context toAttach);

// Returns the context associated with the current scope, will never return null.
static Context current();

// Creates a Context.Key with the given debug name.
static <T> Context.Key<T> key(String name);

// Returns empty context
static Context background();
}
```

The main advantages are:
* Context can be reused in the future by other apis like "tags" etc.
* As you can see the detach requires a toAttach context, this makes behavior deterministic:
```
void foo() {
  Context prev = Context.withValue(...).attach;
  try {
    // Imagine that bar changes the current context without cleaning before return.
    bar();
  } finally {
    // This makes sure that prev will be current after this execution.
    Context.current().detach(prev);
  }
  // here 100% sure prev is the current context.
}
```

### LINKED ISSUE #57: Add support for "detached" spans
Imagine this code in a framework:

```
class InterceptorFoo {
  // executed at the beginning
  void onReceive() {
    Tracer.start("xyz");
  }

  // executed to run the request
  void onRun(Handler h) {
    Tracer.addAnnotation("starting executing handler");
    h();
    Tracer.addAnnotation("finished executing handler");
  }

  // executed at the end of the request
  void onEnd() {
    Tracer.endSpan()
  } 
}
```

This code does not have any way to guarantee that you are appending to the right span or that you end the right span because the framework may also use opencensus and they may call:

```
void on_receive_framework_function() {
  Tracer.start("MyFrameworkSpan");
  try {
    interceptor.onReceive();
  } finally {
    Tracer.endSpan();
  }
}
```

We should support have a way to do something that is more secure here:

```
class InterceptorFoo {
  Span span;
  // executed at the beginning
  void onReceive() {
    span = Tracer.startSpan("xyz");
  }

  // executed to run the request
  void onRun(Handler h) {
    span.addAnnotation("starting executing handler");
    // make span current
    Scope s = Tracer.withSpan(span);
    try {
      h();
    } finally {
      // Just detach the span from the context.
      s.close();
    }
    span.addAnnotation("finished executing handler");
  }

  // executed at the end of the request
  void onEnd() {
    span.end()
  } 
}
```

where `Scope` just attaches/detaches the span from the context

```
interface Scope {
  // Must be called in a try statement in the finally part
  void close();
}
```

With the same interface, now that we have startSpan returning a detached span we can have a "inScope" (not necessary this name):

```
void foo() {
    // starts a span and attaches it to the context
    Scope s = Tracer.inScope("myScopedSpan");
    try {
      doSomethinInSpan();
    } finally {
      // detaches the span from the context and also calls span#end().
      s.close();
    }
}
```

### PR DESCRIPTION
**Pull request by chingor13**
Introduces a way to create detached spans and attach them at a later time. Creates a new `OpenCensus\Core\Scope` class which can be used to close a scope after it's started. Additionally, the `ContextTracer` implementation now uses `Context` and `Scope` for it's implementation.

Fixes #56.
Fixes #57.

### CODE REVIEW

**Comment by tmatsuo** [comment]
Does it make sense to tag a packagist release 0.0.4 which has a strict dependency on ext-opencensus:0.0.4 before making this change?

**Comment by chingor13** [comment]
Sure, can do that. There will also be more extension changes with the upcoming renaming (labels -> attributes)

**Comment by chingor13** [comment]
Oh, it might not be possible as the extension isn't actually a hard requirement.

File: src/Trace/Exporter/StackdriverExporter.php [hunk 1]
```diff
@@ -22,7 +22,8 @@
 use Google\Cloud\Trace\TraceClient;
 use Google\Cloud\Trace\TraceSpan;
 use OpenCensus\Trace\Tracer\TracerInterface;
-use OpenCensus\Trace\Span as OpenCensusSpan;
+use OpenCensus\Trace\Span;
+use OpenCensus\Trace\SpanContext;
```
[the remark this answers is not in the document]
Comment (chingor13) [reply src/Trace/Exporter/StackdriverExporter.php]: At one point it was (a WIP version had `SpanContext::current()` implemented). I'll remove.

File: src/Trace/Exporter/ZipkinExporter.php [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 use OpenCensus\Trace\Tracer\TracerInterface;
 use OpenCensus\Trace\Span;
+use OpenCensus\Trace\SpanContext;
```
[the remark this answers is not in the document]
Comment (chingor13) [reply src/Trace/Exporter/ZipkinExporter.php]: Same as for the `StackdriverExporter` - I'll remove

**Pull request merged**